File: 07/namer.rb
# namer.rb tests class Name and its operations
#
# Begun by: Dr. Adams, for CS 214 at Calvin College.
# Completed by: Charles Blum
# Date: 09 April 2014
####################################################

require 'test/unit/assertions'   # needed for assert
include Test::Unit::Assertions

class Name
  #default constructor
  def initialize(first, middle, last)
    @first, @middle, @last = first, middle, last
  end

  #getters and setters for the instance variables
  attr_accessor :first, :middle, :last

  #fullName creates a string which contains the
  #full name of the person
  def fullName
    myFullName = @first + " " + @middle + " " + @last
  end

  #print prints the full name out
  #and returns the name as a string
  def print
    puts(fullName)
    fullName
  end

  #return a name in last, first (middle initial) format
  def lfmi
    myLFMI = @last + ", " + @first + " " + @middle.chars.first + "."
  end

  #read in a new full name from the user
  def readName
    puts("\nEnter a new full name:\n")
    @first = gets
    @middle = gets
    @last = gets
  end
end

def testName
   name = Name.new("John", "Paul", "Jones")

   assert name.first == "John", "first failed"
   assert name.middle == "Paul", "middle failed"
   assert name.last == "Jones", "last failed"
   assert name.fullName == "John Paul Jones", "fullName failed"
   assert name.print == "John Paul Jones", "print failed"
   print "lab07 tests passed!\n\n"

   name.first, name.middle, name.last  = "Paul", "John", "Smith"
   assert name.first == "Paul", "first failed"
   assert name.middle == "John", "middle failed"
   assert name.last == "Smith", "last failed"
   assert name.lfmi == "Smith, Paul J.", "lfmi failed"
   name.readName
   assert name.first != "Paul", "first failed"
   assert name.middle != "John", "middle failed"
   assert name.last != "Smith", "last failed"
   print "Project07 tests passed!\n"

   print "All tests passed!\n"
end

testName

File: 07/namer.cpp
/* namer.cpp implements and tests class Name.
 *
 * Begun by: Dr. Adams, CS 214 at Calvin College.
 * Completed by:Charles Blum
 * Date: 09 April 2014
 */

#include <iostream>        // istream, ostream
#include <string>          // string
#include <cassert>         // assert()
using namespace std;

// Replace this line with the declaration of class Name 
class Name {
  public:                          // interface functions
    Name(const string & fist, const string & middle, const string & last);
    string getFirst()  const;
    string getMiddle() const;
    string getLast()   const;
    string getFullName() const;
    void print(ostream & out);
    //project07
    void setFirst( const string & f);
    void setMiddle( const string & m);
    void setLast( const string & l );
    string lfmi() const;
    void readName( istream & in );
  private:
    string myFirst,
           myMiddle,
           myLast;
};

/**************************************************
* Name constructs a name from three strings.      *
* Receive: first, middle and last, three strings. *
* Return: the triplet (first middle last).        *
***************************************************/
Name::Name(const string & first, const string & middle, const string & last) {
  myFirst = first;
  myMiddle = middle;
  myLast = last;
}

/***************************************************
 * getFirst() extracts the first name of a Name object. *
 * Return: myFirst.                                *
 ***************************************************/
string Name::getFirst() const {
  return myFirst;
}

/*****************************************************
 * getMiddle() extracts the middle name of a Name object. *
 * Return: myMiddle.                                 *
 *****************************************************/
string Name::getMiddle() const {
  return myMiddle;
}

/**************************************************
 * getLast() extracts the last name of a Name object.  *
 * Return: myLast.                                *
 **************************************************/
string Name::getLast() const {
  return myLast;
}

/***************************************************
 * getFullName() returns a full name in F-M-L order.    *
 * Return: myFirst, myMiddle, myLast.              *
 ***************************************************/
string Name::getFullName() const {
  return myFirst + ' ' + myMiddle + ' ' + myLast;
}

/****************************************************
 * print() displays a Name object.                    *
 * Receive: out, the name of a stream.              *
 * Output: the strings in theName to buffer.        *
 * POST: out contains myFirst, myMiddle and myLast. *
 ****************************************************/
void Name::print(ostream & out) {
  out << getFullName();
}

/*
 * Project 07 Chagnes
 */
/*************************************************
 * Setters for first, middle, and last names.    *
 * Postcondition: the instance variable is set   *
 *************************************************/
void Name::setFirst(const string & aFirst) {
  myFirst = aFirst;
}

void Name::setMiddle(const string & aMiddle) {
  myMiddle = aMiddle;
}

void Name::setLast(const string & aLast) {
  myLast = aLast;
}

/*************************************************
 * lfmi returns the name in last, first middle   *
 * where middle is the first letter.             *
 * Postcondition: name is returned to the user.  *
 *************************************************/
string Name::lfmi() const {
  return myLast + ", " + myFirst + " " + myMiddle[0] + ".";
}

/*************************************************
 * readName takes in new input from the user.    *
 * Postcondition: the instance variables are set *
 * set to new values.                            *
 *************************************************/
void Name::readName(istream & in ) {
  in >> myFirst >> myMiddle >> myLast;
}
/*
 * End Project 07 Changes
 */

/******************************************
 * a simple driver to test our Name type. *
 ******************************************/
int main()
{
  Name aName("John", "Paul", "Jones");

  assert( aName.getFirst() == "John" );
  assert( aName.getMiddle() == "Paul" );
  assert( aName.getLast() == "Jones" );
  assert( aName.getFullName() == "John Paul Jones" );
  
  aName.print(cout);  cout << endl;
  cout << "\nLab07 tests passed!" << endl;

  aName.setFirst( "Paul" );
  aName.setMiddle( "John" );
  aName.setLast( "Smith" );
  
  assert( aName.getFirst() == "Paul" );
  assert( aName.getMiddle() == "John" );
  assert( aName.getLast() == "Smith" );
  assert( aName.getFullName() == "Paul John Smith" );
  cout << aName.lfmi() << endl;
  assert( aName.lfmi() == "Smith, Paul J." );

  cout << "Enter a name that is not: \"" << aName.getFullName() << "\"" << endl;
  aName.readName( cin );

  assert( aName.getFirst() != "Paul" );
  assert( aName.getMiddle() != "John" );
  assert( aName.getLast() != "Smith" );
  assert( aName.getFullName() != "Paul John Smith" );
  assert( aName.lfmi() != "Smith, Paul J." );
  cout << "\nProject07 tests passed!" << endl;

  cout << "\n\nAll tests passed! " << endl;
}

